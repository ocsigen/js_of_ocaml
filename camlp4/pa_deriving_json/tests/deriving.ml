open Deriving_Json

let test name v = Printf.printf "%s = %s\n%!" name v

let () = test "char" (Json_char.to_string '4')

let () = test "bool" (Json_bool.to_string true)

let () = test "bool" (Json_bool.to_string false)

let () = test "unit" (Json_unit.to_string ())

let () = test "int" (Json_int.to_string 42)

let () = test "int32" (Json_int32.to_string 42l)

let () = test "int64" (Json_int64.to_string 42L)

(* let () = test "nativeint" (Json_nativeint.to_string 42n);; *)

let () = test "float" (Json_float.to_string  42.42)

let () = test "string" (Json_string.to_string  "42")

module Int_list   = Json_list(Json_int)

module Int_ref    = Json_ref(Json_int)

module Int_option = Json_option(Json_int)

module Int_array  = Json_array(Json_int)

let () = test "int_list" (Int_list.to_string  [4;2;42])

let () = test "int_ref" (Int_ref.to_string (ref 42))

let () = test "int_option" (Int_option.to_string (Some 42))

let () = test "int_option" (Int_option.to_string None)

let () = test "int_array" (Int_array.to_string [|4;2;42;24|])

type tuple1 = (int * string) deriving (Json)

type variant1 = A | B | C | D of variant1 deriving (Json)

type variant2 = D of string | E of variant1 deriving (Json)

type record1 = { f : variant1; g : variant2; h : record1 option } deriving (Json)

type poly1 = [`A | `B of string] deriving (Json)

type poly2 = [poly1 | `C of int] deriving (Json);;

(* Weird camlp4 issue. *)
();;let () = ()

let () = test "tuple1" (Json_tuple1.to_string (42,"42"))

let () = test "variant1 A" (Json_variant1.to_string A)

let () = test "variant1 B" (Json_variant1.to_string B)

let () = test "variant1 C" (Json_variant1.to_string C)

let () = test "variant2 D 'hello'" (Json_variant2.to_string (D "hello"))

let () = test "variant2 E A" (Json_variant2.to_string (E A))

let () = test "record1" (Json_record1.to_string {f = A; g = D "d"; h = None })

let () = test "record1"
           (Json_record1.to_string
              {f = A;
               g = D "d";
               h = Some {f = B;
                         g = E (D B);
                         h = None }
              })

let () = test "poly1 `A" (Json_poly1.to_string `A)

let () = test "poly1 `B str" (Json_poly1.to_string (`B "str"))

let () = test "poly2 `A" (Json_poly2.to_string `A)

let () = test "poly2 `B str" (Json_poly2.to_string (`B "str"))

let () = test "poly2 `C 42" (Json_poly2.to_string (`C 42));;
