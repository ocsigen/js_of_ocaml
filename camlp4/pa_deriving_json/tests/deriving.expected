

#                                                                                                               char = 52
bool = 1
bool = 0
unit = 0
int = 42
int32 = 42
int64 = [255,42,0,0]
float = 4.242000e+01
string = "42"
int_list = [0,4,[0,2,[0,42,0]]]
int_ref = [0,42]
int_option = [0,42]
int_option = 0
int_array = [0,4,2,42,24]
val test : string -> string -> unit = <fun>
module Int_list :
  sig
    type a = Deriving_Json.Json_int.a list
    val t : a Deriving_Json.t
    val write : Buffer.t -> a -> unit
    val read : Deriving_Json_lexer.lexbuf -> a
    val to_string : a -> string
    val from_string : string -> a
    val match_variant : [ `Cst of int | `NCst of int ] -> bool
    val read_variant :
      Deriving_Json_lexer.lexbuf -> [ `Cst of int | `NCst of int ] -> a
  end
module Int_ref :
  sig
    type a = Deriving_Json.Json_int.a ref
    val t : a Deriving_Json.t
    val write : Buffer.t -> a -> unit
    val read : Deriving_Json_lexer.lexbuf -> a
    val to_string : a -> string
    val from_string : string -> a
    val match_variant : [ `Cst of int | `NCst of int ] -> bool
    val read_variant :
      Deriving_Json_lexer.lexbuf -> [ `Cst of int | `NCst of int ] -> a
  end
module Int_option :
  sig
    type a = Deriving_Json.Json_int.a option
    val t : a Deriving_Json.t
    val write : Buffer.t -> a -> unit
    val read : Deriving_Json_lexer.lexbuf -> a
    val to_string : a -> string
    val from_string : string -> a
    val match_variant : [ `Cst of int | `NCst of int ] -> bool
    val read_variant :
      Deriving_Json_lexer.lexbuf -> [ `Cst of int | `NCst of int ] -> a
  end
module Int_array :
  sig
    type a = Deriving_Json.Json_int.a array
    val t : a Deriving_Json.t
    val write : Buffer.t -> a -> unit
    val read : Deriving_Json_lexer.lexbuf -> a
    val to_string : a -> string
    val from_string : string -> a
    val match_variant : [ `Cst of int | `NCst of int ] -> bool
    val read_variant :
      Deriving_Json_lexer.lexbuf -> [ `Cst of int | `NCst of int ] -> a
  end
type tuple1 = int * string
module Json_tuple1 :
  sig
    type a = tuple1
    val t : a Deriving_Json.t
    val write : Buffer.t -> a -> unit
    val read : Deriving_Json_lexer.lexbuf -> a
    val to_string : a -> string
    val from_string : string -> a
    val match_variant : [ `Cst of int | `NCst of int ] -> bool
    val read_variant :
      Deriving_Json_lexer.lexbuf -> [ `Cst of int | `NCst of int ] -> a
  end
type variant1 = A | B | C | D of variant1
module Json_tdL9PCNdWpVcu7uA6UmEiRfTGowsbaZR :
  sig
    val make_variant1_1 :
      (module Deriving_Json.Json with type a = variant1) Lazy.t
  end
module Json_variant1 :
  sig
    type a = variant1
    val t : a Deriving_Json.t
    val write : Buffer.t -> a -> unit
    val read : Deriving_Json_lexer.lexbuf -> a
    val to_string : a -> string
    val from_string : string -> a
    val match_variant : [ `Cst of int | `NCst of int ] -> bool
    val read_variant :
      Deriving_Json_lexer.lexbuf -> [ `Cst of int | `NCst of int ] -> a
  end
type variant2 = D of string | E of variant1
module Json_variant2 :
  sig
    type a = variant2
    val t : a Deriving_Json.t
    val write : Buffer.t -> a -> unit
    val read : Deriving_Json_lexer.lexbuf -> a
    val to_string : a -> string
    val from_string : string -> a
    val match_variant : [ `Cst of int | `NCst of int ] -> bool
    val read_variant :
      Deriving_Json_lexer.lexbuf -> [ `Cst of int | `NCst of int ] -> a
  end
type record1 = { f : variant1; g : variant2; h : record1 option; }
module Json_XfaHIZqeC7h4bbB6H_eBvyzvX5Drp4_d :
  sig
    val make_record1_1 :
      (module Deriving_Json.Json with type a = record1) Lazy.t
  end
module Json_record1 :
  sig
    type a = record1
    val t : a Deriving_Json.t
    val write : Buffer.t -> a -> unit
    val read : Deriving_Json_lexer.lexbuf -> a
    val to_string : a -> string
    val from_string : string -> a
    val match_variant : [ `Cst of int | `NCst of int ] -> bool
    val read_variant :
      Deriving_Json_lexer.lexbuf -> [ `Cst of int | `NCst of int ] -> a
  end
type poly1 = [ `A | `B of string ]
module Json_poly1 :
  sig
    type a = poly1
    val t : a Deriving_Json.t
    val write : Buffer.t -> a -> unit
    val read : Deriving_Json_lexer.lexbuf -> a
    val to_string : a -> string
    val from_string : string -> a
    val match_variant : [ `Cst of int | `NCst of int ] -> bool
    val read_variant :
      Deriving_Json_lexer.lexbuf -> [ `Cst of int | `NCst of int ] -> a
  end
type poly2 = [ `A | `B of string | `C of int ]
module Json_poly2 :
  sig
    type a = poly2
    val t : a Deriving_Json.t
    val write : Buffer.t -> a -> unit
    val read : Deriving_Json_lexer.lexbuf -> a
    val to_string : a -> string
    val from_string : string -> a
    val match_variant : [ `Cst of int | `NCst of int ] -> bool
    val read_variant :
      Deriving_Json_lexer.lexbuf -> [ `Cst of int | `NCst of int ] -> a
  end
#                                                                   
  
  
  
  
  
  
  
  
tuple1 = [0,42,"42"]
variant1 A = 0
variant1 B = 1
variant1 C = 2
variant2 D 'hello' = [0,"hello"]
variant2 E A = [1,0]
record1 = [0,0,[0,"d"],0]
record1 = [0,0,[0,"d"],[0,[0,1,[1,[0,1]],0]]]
poly1 `A = 65
poly1 `B str = [0,66,"str"]
poly2 `A = 65
poly2 `B str = [0,66,"str"]
poly2 `C 42 = [0,67,42]
# 
